### PR TITLE
tests/main/lxd: add test for snaps inside nested lxd containers not working

### PR DESCRIPTION
**Pull request by anonymouse64**
Trying to run snaps inside a nested LXD container fails thusly:
```
+ lxd.lxc exec my-nesting-ubuntu -- lxd.lxc exec my-inner-ubuntu -- snap install snap install test-snapd-sh
error: system does not fully support snapd: apparmor detected but insufficient
       permissions to use it
```

which is expected because AppArmor only supports one level of nested namespaces, so the nested LXD container would be level 2 and thus is denied. 

Breaking the snapd setup in the container into it's own script makes it easier
to manage this and reduces the duplication.

Also now that 20.04 is released, stop using 19.10 as the 20.04 container.

### CODE REVIEW

File: tests/main/lxd/task.yaml [hunk 1]
```diff
@@ -141,17 +125,34 @@ execute: |
...
+            lxd.lxc exec my-inner-ubuntu -- \
+                sh -c "echo https_proxy=$https_proxy >> /etc/environment"
+    fi
+    lxd.lxc exec my-nesting-ubuntu -- \
+        lxd.lxc exec my-inner-ubuntu -- \
+            /bin/bash /root/prep-snapd-in-lxd.sh
+
+    lxd.lxc exec my-nesting-ubuntu -- \
+        lxd.lxc exec my-inner-ubuntu -- \
+            snap install snap install test-snapd-sh 2>&1 | MATCH error: system does not fully support snapd: apparmor detected but insufficient permissions to use it
```
Comment (stolowski) [inline tests/main/lxd/task.yaml:154]: I think you meant just `snap install test-snapd-sh ...`

File: tests/main/lxd/prep-snapd-in-lxd.sh [hunk 2]
```diff
@@ -0,0 +1,18 @@
+#!/bin/bash -ex
```
Comment (zyga) [inline tests/main/lxd/prep-snapd-in-lxd.sh:1]: Nitpick, use sh when the extension is .sh and we don't need bash features.

File: tests/main/lxd/task.yaml [hunk 3]
```diff
@@ -70,16 +70,8 @@ execute: |
     # prep two containers, the my-ubuntu normal container and the 
     # my-nesting-ubuntu nesting container
 
-    (
-        . /etc/os-release
-        # TODO:2004: remove version check once the image for ubuntu 20.04 is published
```
Comment (zyga) [inline tests/main/lxd/task.yaml:75]: Nice :)

File: tests/main/lxd/task.yaml (hunk 1 above)
Comment (anonymouse64) [reply tests/main/lxd/task.yaml:154]: /o\ yes indeed haha

File: tests/main/lxd/prep-snapd-in-lxd.sh (hunk 2 above)
Comment (anonymouse64) [reply tests/main/lxd/prep-snapd-in-lxd.sh:1]: sure

File: tests/main/lxd/task.yaml [hunk 4]
```diff
@@ -149,9 +149,10 @@ execute: |
         lxd.lxc exec my-inner-ubuntu -- \
             /bin/bash /root/prep-snapd-in-lxd.sh
 
-    lxd.lxc exec my-nesting-ubuntu -- \
+    not lxd.lxc exec my-nesting-ubuntu -- \
         lxd.lxc exec my-inner-ubuntu -- \
-            snap install snap install test-snapd-sh 2>&1 | MATCH error: system does not fully support snapd: apparmor detected but insufficient permissions to use it
+            snap install test-snapd-sh 2>stderr.log
+    cat stderr.log | MATCH error: system does not fully support snapd: apparmor detected but insufficient permissions to use it
```
Comment (stolowski) [inline tests/main/lxd/task.yaml:155]: Nitpick, this could be simplified to `MATCH ... < stderr.log`.
Also, shouldn't the text to match be quoted?

Comment (anonymouse64) [reply tests/main/lxd/task.yaml:155]: ah yes you're exactly right thanks for catching this

**Comment by mvo5** [comment]
This fails in the 20.04 test:
```
...
2020-05-04T20:02:11.5794839Z + not lxd.lxc exec my-nesting-ubuntu -- lxd.lxc exec my-inner-ubuntu -- snap install test-snapd-sh
2020-05-04T20:02:11.5795198Z + MATCH 'error: system does not fully support snapd: apparmor detected but insufficient\s+permissions to use it'
2020-05-04T20:02:11.5795332Z grep error: pattern not found, got:
2020-05-04T20:02:11.5795457Z error: system does not fully support snapd: apparmor detected but insufficient
2020-05-04T20:02:11.5795574Z        permissions to use it
2020-05-04T20:02:11.5795801Z -----
2020-05-04T20:02:11.5795905Z .
2020-05-04T20:02:13.1377187Z 2020-05-04 20:02:13 Debug output for google:ubuntu-20.04-64:tests/main/lxd (may041949-462322) : 
...
```

**Comment by zyga** [comment]
@anonymouse64 I think there's something that controls wrapping of our messages, we should perhaps disable that, or set it to a larger value to avoid this sort of wrapping problems.

**Comment by anonymouse64** [comment]
Regex is hard :-(

**Comment by anonymouse64** [comment]
Okay finally fixed the issue here by replacing all newlines in stderr with spaces and then regex matching any amount of whitespace between words, so that if the columns changes etc and we have different line splits, it will be robust. :tada:

**Pull request merged**